fields/types/gcsavatar/GcsAvatarType.js

```diff
@@ -0,0 +1,401 @@
+  });
+
+  const reset = function(item) {
+    item.set(_this.path, {
+      filename: '',
+      filetype: '',
+      gcsBucket: '',
+      gcsDir: '',
+    });
+  };
+
+  const schemaMethods = {
+    exists: function() {
+      return exists(this);
+    },
+    /**
+     * Resets the value of the field
+     *
+     * @api public
+     */
+    reset: function() {
+      reset(this);
+    },
+    /**
+     * Deletes the file from gcs and resets the field
+     *
+     * @api public
+     */
+    delete: function() {
+      const _this = this;
+      const promise = new Promise(function(resolve, reject) {
+        const gcsConfig = keystone.get('gcs config');
+        const bucket = gcsHelper.initBucket(gcsConfig, _this.get(paths.gcsBucket));
+        const filename = _this.get(paths.filename);
+        if (filename && typeof filename === 'string') {
+          bucket.deleteFiles({
+            prefix: _this.get(paths.gcsDir) + filename,
+          }, function(err) {
+            if (err) {
+              return reject(err);
+            }
+            resolve();
+          });
+        } else {
+          resolve();
+        }
+      });
+      reset(this);
+      return promise;
+    },
+  };
+
+  _.forEach(schemaMethods, function(fn, key) {
+    _this.underscoreMethod(key, fn);
+  });
+
+  // expose a method on the field to call schema methods
+  this.apply = function(item, method) {
+    return schemaMethods[method].apply(item, Array.prototype.slice.call(arguments, 2));
+  };
+
+  this.bindUnderscoreMethods();
+};
+
+
+/**
+ * Formats the field value
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.format = function(item) {
+  if (this.hasFormatter()) {
+    return this.options.format(item, item[this.path]);
+  }
+  return item.get(this.paths.filename);
+};
+
+
+/**
+ * Detects the field have formatter function
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.hasFormatter = function() {
+  return typeof this.options.format === 'function';
+};
+
+
+/**
+ * Detects whether the field has been modified
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.isModified = function(item) {
+  return item.isModified(this.paths.filename);
+};
+
+
+/**
+ * Validates that a value for this field has been provided in a data object
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.inputIsValid = function(data) {
+  // TODO - how should file field input be validated?
+  return true;
+};
+
+
+/**
+ * Updates the value for this field in the item from a data object
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.updateItem = function(item, data, callback) {
+  // TODO - direct updating of data (not via upload)
+  process.nextTick(callback);
+};
+
+/**
+ * Uploads the file for this field
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.uploadFile = function(item, file, update, callback) {
+  const _this = this;
+  const ONE_YEAR = 60 * 60 * 24 * 365;
+  const publicRead = _this.options.publicRead ? _this.options.publicRead : false;
+  let filename = file.name;
```

Comment:
I think this line could be omitted since we would always rewrite it at Ln245.

Reply:
Updated.